jv2dmk.man
.TH jv2dmk 1
.SH Name
jv2dmk \- Convert a JV1 or JV3 floppy disk image file to DMK format
.SH Syntax
.B jv2dmk [options] filename.dsk [filename.dmk]
.SH Description
The jv2dmk program converts a floppy disk image in the JV1 or JV3
format to the DMK format.  All of these formats were originally
developed for TRS-80 Model I/III/4/4P emulators, though the DMK format
is useful more broadly.

For more information about the DMK and JV3 file formats and the
emulators that use them, see:
.nf
    http://discover-net.net/~dmkeil/
    http://www.tim-mann.org/xtrs.html
.fi
.SH Options
If only filename.dsk is given on the command line, filename.dmk is
formed by deleting any extension from filename.dsk and appending ".dmk".
.TP
.B \-v \fIverbosity\fP
Specify how much output is printed.  Larger numbers select more output.
The default setting is 0.
.RS
.TP
0
No output.
.TP
1
Print a line for each track, listing the sector numbers found.
.TP
2
Print a line of data about each sector.
.RE
.TP
.B \-j \fIwhich\fP
Specify whether the input disk image is JV1 (-j1) or JV3 (-j3).  
The default is 3.
.TP
.B \-k \fIkind\fP
Specify the type of media in use.  The default is 1.  Possible values:
.RS
.TP
1
5.25-inch or 3.5-inch SD/DD disk
.TP
2
Same as 1; accepted for compatibility with cw2dmk.
.TP
3
5.25-inch HD disk or 8-inch SD/DD disk
.TP
4
3.5-inch HD disk
.RE
.TP
.B \-i \fIiamark\fP
If this option is set to 1, jv2dmk writes an IBM gap0 and 
index address mark at the start of each track.  If the option is set to 0
(the default), jv2dmk uses an ISO format with no gap0 or
index address mark.
With very tightly-packed formats, setting the -i1 option could cause
a track to be too long to fit on the disk.
.TP
.B \-w \fIfmtimes\fP
Normally, FM bytes are written into the DMK file twice (-w2),
so that they take up the correct proportion of the space on mixed-density
tracks.  You can set -w1 to cause FM bytes to be written only once.
This does not save space in the DMK file unless you also reduce the track
length with the -l option.
.TP
.B \-l \fIbytes\fP
DMK track length in bytes, overriding the default setting established
by the -k option.  This option is not effective unless given after 
the -k option.
.SH Limitations
It would be nice if jv2dmk could guess the right value for the -k flag.
Although the JV3 format doesn't record this value, one could probably
do a good job of guessing by using the smallest -k for which the JV3's
track 0 data fits.  A JV1 image is always -k1.

In rare cases a JV3 can be unrepresentable in DMK format.  This
happens if the JV3 has too much data or too many sectors on one
track.  A real copy-protected disk can be written with sectors that
overlap, making it appear when the disk is converted to JV3 that there
is more data on the track than is physically possible, and/or that
there are more sectors on the track than the DMK limit of 64.

On extremely tightly-packed formats (unlikely to occur in practice),
jv2dmk's gap computation might fail to produce usable results.  It
might complain that the track is too long when a clever human could
have made it fit, or worse, it might make gap3 (the gap immediately
after a sector's data and CRC) too small to allow for safely
overwriting a sector later.  If gap3 is too small, overwriting one
sector can destroy the sector ID of the next sector.

jv2dmk does not support the xtrs-only "non-IBM" extension to JV3 format.
.SH Diagnostics
.TP 
.B jv2dmk: Too many sectors per track (n)
The DMK format has a limit of 64 sectors per track.
.TP
.B jv2dmk: Physical track too long by n bytes
The JV3 file had too much data on the track to fit on the specified
kind of disk.  You may have used the wrong value for the -k flag, or
jv2dmk may not have been clever enough to work out how to fit the
track; see Limitations above.
.TP
.B jv2dmk: DMK track too long by n bytes
This error should occur only if you gave the -l flag with too small a value.
.SH Authors
jv2dmk was written by Timothy Mann, http://tim-mann.org/.
jv2dmk is free software, released under the GNU General Public License.
Thanks to David Keil for designing and
documenting the DMK file format for floppy disk images,
and to Jeff Vavasour for designing and documenting the JV3 format.


ChangeLog
4.9.1 -- Mon May 23 15:56:42 UTC 2022 -- Tim Mann

* Add -T step_time,settling_time feature to both cw2dmk and dmk2cw.

* Log all Catweasel samples when at verbosity level 7, instead of
  stopping when the DMK buffer is full.  (This change was actually
  made in 4.9.0, but it had a bug that is now fixed.)

4.9.0 -- Sat Apr 30 17:07:57 UTC 2022 -- Tim Mann, Quentin Barnes

* Catch errors when writing to files.   Report and exit when errors
  are seen.

* Add -X min_retry feature to cw2dmk, accepting the same syntax for
  min_retry as for -x max_retry.

* Add QUIRK_IAM to -q option in cw2dmk.

* Various minor cleanup of code and the cw2dmk usage message.

* Add -R (replay) feature to cw2dmk.

* Add QUIRK_EXTRA_DATA to -q option in cw2dmk.

4.8.1 -- Sat Dec 11 01:12:14 UTC 2021 -- Quentin Barnes

* Correct a substantial performance regression when doing high
  level writes to the log file.

4.8.0 -- Tue Dec  7 20:52:50 UTC 2021 -- Quentin Barnes

* This release is virtually all of Tim Mann's work, and most of
  those changes are for adding UDOS support.

* Add -q (quirk) option for extending cw2dmk to better support of
  reading UDOS formatted disks.  UDOS disks may have a modified
  ID premark and after each data sector may include an extra 4 data
  bytes and a possible 2 bytes of CRC.  For details, see the cw2dmk
  man page's description of the -q option.

  The new -q option hasn't been tested with the -j option.  Bugs
  may show up.  If found, please report them.

* Add -a rate_adj and -e dither options to dmk2cw.

4.7.1 -- Wed Oct 20 20:00:21 UTC 2021 -- Quentin Barnes

* Add feature to "-x" option to allow per-track retry counts.
  (E.g. "-x 40,0:4-12" sets retries for all tracks to 40 except 
  disable retries for tracks 4 through 12.)  See cw2dmk man
  page for further information.

* Correct off-by-one problem with retry counts.

4.7.0 -- Mon Apr 12 23:25:25 UTC 2021 -- Quentin Barnes

* Deprecate cw2dmk's -b option introduced in 4.6.0.  It had too
  many shortcomings with CP/M disks.  Replaced with -C.

* Add -C option to cw2dmk for controlling detection of incompatibly
  formatted sides of a floppy disk.  If side 1 is formatted with
  512-byte sectors and side 0 is formatted with any other sector
  size, it'll assume side 1 was factory-formatted and skip it.
  Detection of other incompatible combinations may be added in a
  future release.  Use "-C 0" to disable the detection.

* Add new interactive menu.  When enabled with -Mi and the interrupt
  key is pressed (typically ^C), cw2dmk will enter a menu with
  options.  Also, -Me will enable dropping into the menu when the
  error limit for a track exceeds the number set by -x.

4.6.0 -- Thu Dec 17 17:24:34 UTC 2020 -- Quentin Barnes

* Skipping over 4.5.x versions since George Phillips had used that
  point release line for his work.

* Git repo "cw2dmk" of Tim Mann's work was created by ezrec
  (Jason S. McMullan).  Transferred to qbarnes.

* Switched Linux release binaries to be for 64-bit OSes.

* Added change from George Phillips for his "join" (-j) feature
  that will join different rereads of the same track and merge
  good sectors.

* Fixed problem of missing data between index hole detect and read
  started.  This should noticably reduce disk read errors and even
  occasionally dropping of entire sectors.

* Added change to allow cw2dmk tools to function on newer Linux
  releases by enabling PCI I/O space access if disabled.

* Added change to recognize and skip factory-formatted sides of a
  disk.  Controlled with -b option.

* The subdirectory "cwsdpmi" is now a git sub-module for picking up
  the "cwsdpmi.exe" binary for MS-DOS.

* Cleaned up and merged multiple makefiles for different environments
  into one.  Makes it easier to maintain, build, and cross-build.  By
  default, the Makefile will now only build for the native OS.  See
  the new file BUILDING.md for further information.

* Created subdirectory "firmware" for holding firmware related files
  including a LICENSE file.  I also received some additional
  documentation from Jens Schoenfeld on programming the MK3 and MK4
  and a file with firmware revision history saved here as
  firmware/Changelog.

* To be less generic, renamed "testhist" binary to "cwtsthst", still
  keeping to MS-DOS's 8.3 file naming convention.

* Update code base to be 64-bit clean.  Fixed all known problems but
  some may remain.

* Changed to using semantic versioning for version ID.

4.4 -- Sat Nov 28 11:45:01 PST 2009 -- Tim Mann

* The Linux binaries included in this distribution are compiled on a
  32-bit Ubuntu 8.04.3 LTS system.  The MS-DOS binaries are still
  compiled with a version of djgpp from the year 2000, under djgpp.
  The MS-DOS binaries are untested at the time of this release.
	
* Updated to the current Catweasel MK4 firmware from Individual
  Computers, moving from pre29.cw4 to rel2f2.cw4.  Changed catweasl.c
  slightly to work with the new firmware -- the write pulse width now
  must be set.

* Fixed cw2dmk bug where the total sectors read in each encoding
  counted all retries, not just the last one.
	
* Added -r option to cw2dmk and dmk2cw to reverse the sides.  This is
  mostly to help split/join disks that have a separate one-sided image
  on each side but were written with a double-sided drive (that is,
  they are not flippy) to/from two separate DMK images.
	
* Fixed an incompatibility with newer versions of libpci that had
  prevented recompiled Linux binaries from recognizing a PCI
  Catweasel.

* Added a dmk2cw option to use only a prefix of the track in the DMK
  file.

* Changed the way dmk2cw generates fill data when the DMK track is
  shorter than needed to fill the disk.  Added an option to select
  some alternative fill methods.

* Some small fixes and cleanups, mostly in documentation and messages.

4.3 -- Sat Jun 25 00:21:23 PDT 2005 -- Tim Mann

* By popular demand, print more stuff at the default verbosity levels.

jv2dmk.c
/*
 * jv2dmk: Convert a JV1 or JV3 format emulated floppy to DMK format.
 * Copyright (C) 2002 Timothy Mann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

#include "dmk.h"
#include "jv3.h"
#include "crc.c"

/* Command line options */
#define OUT_QUIET 0
#define OUT_TRACKS 1
#define OUT_SECTORS 2
int verbosity = OUT_QUIET;
int kind = 1;
int indexmark = 0;
int fmtimes = 2;
int dmklen = 0;
int jvwhat = 3;

/* Kinds */
#define NKINDS 4
typedef struct {
  char *name;
  int nomlen;
  int dmklen;
} KindData;

KindData kindData[] = {
  { "5.25-inch or 3.5-inch SD/DD disk", TRKSIZE_DD, DMK_TRACKLEN_5 },
  { "5.25-inch or 3.5-inch SD/DD disk", TRKSIZE_DD, DMK_TRACKLEN_5 },
  { "5.25-inch HD or 8-inch SD/DD disk", TRKSIZE_5HD, DMK_TRACKLEN_8 },
  { "3.5-inch HD disk", TRKSIZE_3HD, DMK_TRACKLEN_3HD }
};

typedef struct {
  unsigned char track;
  unsigned char sector;
  unsigned char flags;
  off_t offset;
  int sizecode;
} SectorId;

unsigned char damtab[] = { 0xfb, 0xfa, 0xf9, 0xf8, 0xfb, 0xf8, 0xfb, 0xf8 };
#define JV3_GETDAM(flags) (damtab[((flags) & (JV3_DENSITY|JV3_DAM)) >> 5])

unsigned char jv1_interleave[] = { 0, 5, 1, 6, 2, 7, 3, 8, 4, 9 };

SectorId idbuf[JV3_SECSMAX];

int
idcompare(const void* id1v, const void* id2v)
{
#define id1 ((SectorId*)id1v)
#define id2 ((SectorId*)id2v)
  if (id1->track < id2->track) return -1;
  if (id1->track > id2->track) return 1;
  if ((id1->flags & JV3_SIDE) < (id2->flags & JV3_SIDE)) return -1;
  if ((id1->flags & JV3_SIDE) > (id2->flags & JV3_SIDE)) return 1;
  if (id1->offset < id2->offset) return -1;
  if (id1->offset > id2->offset) return 1;
  return 0;
}

int
readidblock(FILE* fin, SectorId** curidp, off_t* offsetp)
{
  int i, c;
  SectorId* curid = *curidp;
  off_t offset = *offsetp;

  for (i=0; i<JV3_SECSPERBLK; i++) {
    c = getc(fin);
    if (c == EOF) return EOF;
    curid->track = c;
    curid->sector = getc(fin);
    curid->flags = getc(fin);
    curid->offset = offset;
    if (curid->track == JV3_FREE) {
      curid->sizecode = (curid->flags & JV3_SIZE) ^ 2;
    } else {
      curid->sizecode = (curid->flags & JV3_SIZE) ^ 1;
    }
    offset += 128 << curid->sizecode;
    curid++;
  }
  *curidp = curid;
  *offsetp = offset;
  return 0;
}

void
dmkput(unsigned char** dmkp, unsigned char byte, int count, int curden)
{
  if (!curden) count *= fmtimes;
  while (count--) *(*dmkp)++ = byte;
}

void
dotrack(FILE* fin, SectorId** curidp, int track, int side,
	int nomlen, int dmklen, unsigned char* dmkbuf)
{
  SectorId* curid = *curidp;
  int secs = 0;
  int lenleft = nomlen;
  int gap1, gap3, gap4;
  unsigned short* idamp = (unsigned short*) dmkbuf;
  unsigned char* dmkp = dmkbuf + DMK_TKHDR_SIZE;
  int curden, i;
  unsigned short crc;
  unsigned char dam;

  /* Count sectors on track and measure minimum size */
  if (verbosity == OUT_TRACKS) printf("track %2d side %d: ", track, side);
  while (curid->track == track &&
	 ((curid->flags & JV3_SIDE) ? 1 : 0) == side) {
    if (verbosity == OUT_TRACKS) {
      printf("%d ", curid->sector);
    } else if (verbosity == OUT_SECTORS) {
      printf("track %2d, side %d, sector %2d: size %d, %cden, dam %02x",
	     track, side, curid->sector, 128 << curid->sizecode,
	     (curid->flags & JV3_DENSITY) ? 'd' : 's',
	     JV3_GETDAM(curid->flags));
      if (curid->flags & JV3_ERROR) printf(", crc error");
      if (curid->flags & JV3_NONIBM) printf(", non-ibm [not supported]");
      printf("\n");
    }

    if (curid->flags & JV3_DENSITY) {
      lenleft -= (128 << curid->sizecode) + 62;
    } else {
      lenleft -= ((128 << curid->sizecode) + 33) * 2;
    }
    secs++;
    curid++;
  }
  if (verbosity == OUT_TRACKS) printf("\n");
  if (secs > DMK_TKHDR_SIZE/2) {
    fprintf(stderr, "jv2dmk: Too many sectors (%d) on track %d\n",
	    secs, track);
    exit(1);
  }

  curid = *curidp;
  curden = (curid->flags & JV3_DENSITY) != 0;
  if (indexmark) {
    /* Write IBM gap 0, index mark, and gap 1 */
    if (curden) {
      dmkput(&dmkp, 0x4e, 80, curden);
      dmkput(&dmkp, 0x00, 12, curden);
      dmkput(&dmkp, 0xc2, 3, curden);
      dmkput(&dmkp, 0xfc, 1, curden);
      dmkput(&dmkp, 0x4e, 50, curden);
    } else {
      dmkput(&dmkp, 0xff, 40, curden);
      dmkput(&dmkp, 0x00, 6, curden);
      dmkput(&dmkp, 0xfc, 1, curden);
      dmkput(&dmkp, 0xff, 26, curden);
    }
  } else {
    /* Write ISO gap 1 */
    if (curden) {
      dmkput(&dmkp, 0x4e, 32, curden);
    } else {
      dmkput(&dmkp, 0xff, 16, curden);
    }
  }
  gap1 = (dmkp - (dmkbuf + DMK_TKHDR_SIZE));

  /* Set gap4 to 1% of the total length to allow for 1%
   * drive speed variation. */
  gap4 = nomlen / 100;

  /* Divide remaining space into gap3's */
  lenleft -= gap1 + gap4;
  if (lenleft < 0) {
    fprintf(stderr, "jv2dmk: Physical track %d too long by %d bytes\n",
	    track, -lenleft);
    exit(1);
  }
  gap3 = lenleft / secs;

  while (secs--) {
    /* Write sync and IDAM */
    if (curden) {
      dmkput(&dmkp, 0x00, 12, curden);
      dmkput(&dmkp, 0xa1, 3, curden);
      crc = calc_crc1(0x968b, 0xa1); /* CRC of a1 a1 a1 */
      *idamp++ = (dmkp - dmkbuf) | DMK_DDEN_FLAG;
    } else {
      dmkput(&dmkp, 0x00, 6, curden);
      crc = 0xffff;
      *idamp++ = dmkp - dmkbuf;
    }
    dmkput(&dmkp, 0xfe, 1, curden);
    crc = calc_crc1(crc, 0xfe);

    /* Write ID */
    dmkput(&dmkp, track, 1, curden);
    crc = calc_crc1(crc, track);
    dmkput(&dmkp, side, 1, curden);
    crc = calc_crc1(crc, side);
    dmkput(&dmkp, curid->sector, 1, curden);
    crc = calc_crc1(crc, curid->sector);
    dmkput(&dmkp, curid->sizecode, 1, curden);
    crc = calc_crc1(crc, curid->sizecode);
    dmkput(&dmkp, crc >> 8, 1, curden);
    crc = calc_crc1(crc, crc >> 8);
    dmkput(&dmkp, crc >> 8, 1, curden);

    /* Write gap 2, sync, and DAM */
    if (curden) {
      dmkput(&dmkp, 0x4e, 22, curden);
      dmkput(&dmkp, 0x00, 12, curden);
      dmkput(&dmkp, 0xa1, 3, curden);
      crc = calc_crc1(0x968b, 0xa1); /* CRC of a1 a1 a1 */
    } else {
      dmkput(&dmkp, 0xff, 11, curden);
      dmkput(&dmkp, 0x00, 6, curden);
      crc = 0xffff;
    }
    dam = JV3_GETDAM(curid->flags);
    dmkput(&dmkp, dam, 1, curden);
    crc = calc_crc1(crc, dam);

    /* Write data */
    fseek(fin, curid->offset, SEEK_SET);
    i = 128 << curid->sizecode;
    while (i--) {
      unsigned char c = getc(fin);
      dmkput(&dmkp, c, 1, curden);
      crc = calc_crc1(crc, c);
    }

    /* Write data CRC */
    dmkput(&dmkp, crc >> 8, 1, curden);
    crc = calc_crc1(crc, crc >> 8);
    if (curid->flags & JV3_ERROR) crc = ~crc;
    dmkput(&dmkp, crc >> 8, 1, curden);

    /* Write gap 3.  Use density of next sector, if any */
    curid++;
    if (secs > 0) {
      curden = (curid->flags & JV3_DENSITY) != 0;
    }
    if (curden) {
      dmkput(&dmkp, 0x4e, gap3, curden);
    } else {
      dmkput(&dmkp, 0xff, gap3/2, curden);
    }
  }

  /* Write gap 4 */
  lenleft = dmklen - (dmkp - dmkbuf);
  if (lenleft < 0) {
    /* This can happen if the user specified too short a dmklength */
    fprintf(stderr, "jv2dmk: DMK track too long by %d bytes\n", -lenleft);
    exit(1);
  }
  dmkput(&dmkp, curden ? 0x4e : 0xff, lenleft, 1);

  *curidp = curid;
}

void
usage()
{
  printf("Usage: jv2dmk [options] file.dsk [file.dmk]\n");
  printf(" Options [defaults in brackets]:\n");
  printf(" -v verbosity  Amount of output [%d]\n", verbosity);
  printf("               0 = No output\n");
  printf("               1 = List of sectors on each track\n");
  printf("               2 = Details about each sector\n");
  printf(" -j which      Specify JV1 or JV3 [%d]\n", jvwhat);
  printf(" -k kind       Specify the type of media [%d]\n", kind);
  printf("               1 or 2 = %s\n", kindData[0].name);
  printf("               3 = %s\n", kindData[2].name);
  printf("               4 = %s\n", kindData[3].name);
  printf(" -i iam        1 = write IBM index address mark; 0 = don't [%d]\n",
	 indexmark);
  printf(" -w fmtimes    Write FM bytes 1 or 2 times [%d]\n", fmtimes);
  printf(" -l bytes      DMK track length in bytes [depends on kind]\n");
  exit(2);
}

int
main(int argc, char *argv[])
{
  char* jv_name = NULL;
  char* dmk_name = NULL;
  FILE* fin;
  FILE* fout;
  SectorId* curid;
  off_t offset;
  int i, totalsectors;
  int maxtrack, maxside, track, side;
  unsigned char *dmkbuf;
  dmk_header_t *dmkheader;
  int nomlen, buflen;
  int ch;

  opterr = 0;
  for (;;) {
    ch = getopt(argc, argv, "v:j:k:i:w:l:");
    if (ch == -1) break;
    switch (ch) {
    case 'v':
      verbosity = strtol(optarg, NULL, 0);
      if (verbosity < OUT_QUIET || verbosity > OUT_SECTORS) usage();
      break;
    case 'j':
      jvwhat = strtol(optarg, NULL, 0);
      if (jvwhat != 1 && jvwhat != 3) usage();
      break;
    case 'k':
      kind = strtol(optarg, NULL, 0);
      if (kind < 1 || kind > NKINDS) usage();
      break;
    case 'i':
      indexmark = strtol(optarg, NULL, 0);
      if (indexmark < 0 || indexmark > 1) usage();
      break;
    case 'w':
      fmtimes = strtol(optarg, NULL, 0);
      if (fmtimes < 1 || fmtimes > 2) usage();
      break;
    case 'l':
      dmklen = strtol(optarg, NULL, 0);
      if (dmklen < 0 || dmklen > 0x4000) usage();
      break;
    default:
      usage();
      break;
    }
  }

  switch (argc - optind) {
  case 2:
    /* Two filenames given */
    jv_name = argv[optind];
    dmk_name = argv[optind+1];
    break;

  case 1: {
    char *p;
    int len;

    jv_name = argv[optind];
    p = strrchr(jv_name, '.');
    if (p == NULL) {
      len = strlen(jv_name);
    } else {
      len = p - jv_name;
    }
    dmk_name = (char *) malloc(len + 5);
    sprintf(dmk_name, "%.*s.dmk", len, jv_name);
    break; }

  default:
    usage();
  }

  nomlen = kindData[kind-1].nomlen;
  if (dmklen == 0) dmklen = kindData[kind-1].dmklen;
  buflen = nomlen + DMK_TKHDR_SIZE;
  if (dmklen > buflen) buflen = dmklen;
  dmkbuf = (unsigned char*) calloc(1, buflen);
  dmkheader = (dmk_header_t *) calloc(1, sizeof(dmk_header_t));

  fin = fopen(jv_name, "rb");
  if (fin == NULL) {
    perror(jv_name);
    exit(1);
  }	
  fout = fopen(dmk_name, "w+b");
  if (fout == NULL) {
    perror(dmk_name);
    exit(1);
  }	

  if (jvwhat == 1) {
    /* Fake up some ids */
    fseek(fin, 0, SEEK_END);
    totalsectors = ftell(fin) / 256;
    maxside = 0;
    maxtrack = (totalsectors + 9) / 10 - 1;
    offset = 0;
    curid = idbuf;
    for (track=0; track<=maxtrack; track++) {
      for (i=0; i<10; i++) {
	curid->track = track;
	curid->sector = jv1_interleave[(i + 9*track) % 10];
	curid->flags = (track == 17) ? JV3_DAMSDFA : 0;
	curid->offset = offset + 256 * curid->sector;
	curid->sizecode = 1;
	curid++;
      }
      offset += 2560;
    }
    curid->track = curid->sector = curid->flags = JV3_FREE;
  } else {
    /* Read up to two blocks of ids */
    curid = idbuf;
    offset = JV3_SECSTART;
    for (i=0; i<2; i++) {
      if (readidblock(fin, &curid, &offset)) break;
      fseek(fin, offset, SEEK_SET);
      offset += JV3_SECSTART;
    }
    totalsectors = curid - idbuf;

    /* Sort the ids by track, side, offset. */
    qsort(idbuf, totalsectors, sizeof(SectorId), idcompare);

    /* Find number of tracks and sides */
    curid = idbuf;
    maxtrack = -1;
    maxside = 0;
    while (curid->track != JV3_FREE) {
      if (curid->track > maxtrack) maxtrack = curid->track;
      if (curid->flags & JV3_SIDE) maxside = 1;
      curid++;
    }
  }

  /* Write DMK header */
  dmkheader->ntracks = maxtrack + 1;
  dmkheader->tracklen = dmklen; 
  dmkheader->options = ((maxside == 0) ? DMK_SSIDE_OPT : 0)
                     + ((fmtimes == 1) ? DMK_SDEN_OPT : 0);
  /* assumes this machine is little-endian: */
  fwrite(dmkheader, sizeof(dmk_header_t), 1, fout);

  /* Write tracks */
  curid = idbuf;
  for (track=0; track<=maxtrack; track++) {
    for (side=0; side<=maxside; side++) {
      dotrack(fin, &curid, track, side, nomlen, dmklen, dmkbuf);
      fwrite(dmkbuf, dmklen, 1, fout);
    }
  }

  return 0;
}
